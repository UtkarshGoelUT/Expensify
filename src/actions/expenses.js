import database from '../firebase/firebase';

const addExpense = (expense) => {
    return {
        type: 'ADD_EXPENSE',
        expense
    }
};

const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const { description = '', note = '', amount = 0, createdAt = 0 } = expenseData;
        const expenses = {
            description, note, amount, createdAt
        }
        const uid = getState().auth.uid;
        database.ref(`users/${uid}/expenses`).push(expenses).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expenses
            }));
        });
    }
}

const removeExpense = ({ id } = {}) => (
    {
        type: 'REMOVE_EXPENSE',
        id
    }
);

export const startRemoveExpense = ({ id } = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        database.ref(`users/${uid}/expenses/${id}`).remove()
            .then(() => {
                dispatch(removeExpense({ id }));
            });
    }
}

const editExpense = (id, updates) => (
    {
        type: 'EDIT_EXPENSE',
        id,
        updates
    }
);

export const startEditExpense = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).update(updates)
            .then(() => {
                dispatch(editExpense(id, updates));
            });
    }
}

export const setExpense = (expenses) => ({
    type: "SET_EXPENSES",
    expenses
})

export const startSetExpense = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses`).once('value')
            .then((snapshot) => {
                const expenses = [];
                snapshot.forEach((childSnapshot) => {
                    expenses.push({
                        id: childSnapshot.key,
                        ...childSnapshot.val()
                    });
                });
                dispatch(setExpense(expenses));
            });
    };
}

export { addExpense, removeExpense, editExpense, startAddExpense };