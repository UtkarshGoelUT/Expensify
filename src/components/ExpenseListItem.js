import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseItemList = (props) => (
    <Link className="list-item" to={`/edit/${props.id}`}>
        <div>
            <h3 className="list-item__title">
                {props.description}
            </h3>
            <span className="list-item__subtitle">
                â‚¹{numeral(props.amount / 100).format('0,0.00')}
            </span>
        </div>
        <h3 className="list-item__data">
            {moment(props.createdAt).format('MMMM Do, YYYY')}
        </h3>
    </Link>
);

export default ExpenseItemList;