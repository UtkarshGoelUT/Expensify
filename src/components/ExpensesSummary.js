import React from 'react';
import { connect } from 'react-redux';
import getVisibleExpenses from '../selectors/expenses';
import expensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpensesSummary = ({ expenses, filters }) => {
    const visible = getVisibleExpenses(expenses, filters);
    if (visible.length > 1) {
        return <div>
            Viewing {visible.length} expenses : Totalling ₹{numeral(expensesTotal(visible) / 100).format('0,0.00')}
        </div>
    }
    else {
        return <div>
            Viewing {visible.length} expense : Totalling ₹{numeral(expensesTotal(visible) / 100).format('0,0.00')}
        </div>
    }
};

const mapStateToProps = (state) => (
    {
        expenses: state.expenses,
        filters: state.filters
    }
)

export default connect(mapStateToProps)(ExpensesSummary);