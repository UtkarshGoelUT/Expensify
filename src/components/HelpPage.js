import React from 'react';

const HelpPage = () => (
    <div>
        <p>
            Ask for help
        </p>
    </div>
);

export default HelpPage;