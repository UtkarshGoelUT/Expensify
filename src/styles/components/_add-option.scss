.add-option-error{
    color: $off-white;
    font-style: italic;
    padding: 0 $m-size;
    margin-top: $m-size;
}

.add-option{
    display:flex;
    flex-direction: column;
    padding: $m-size;
}

.add-option__input{
    background: $dark-blue;
    border: none;
    border-bottom: .3rem solid darken($dark-blue, 10%);
    flex-grow: 1;
    color: $off-white;
    padding: $s-size;
    margin: 0 0 $s-size 0;
}

@media (min-width: $desktop-break-point) {
   .add-option{
       flex-direction: row;
   }
   
   .add-option__input{
        margin-right: $s-size;
   }
}