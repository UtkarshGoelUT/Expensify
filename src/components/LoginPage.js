import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

const LoginPage = ({ startLogin }) => (
    <div className="box-layout">
        <div className="box-layout__box">
            <h1 className="box-layout__title">Expensify</h1>
            <p>Keep track of all your expenses with Expensify</p>
            <button onClick={startLogin} className="button">Login with Google</button>
        </div>

    </div>
);

const mapDispathToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin())
})

export default connect(undefined, mapDispathToProps)(LoginPage);