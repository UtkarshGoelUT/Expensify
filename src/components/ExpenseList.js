import React from 'react';
import ExpenseListItem from './ExpenseListItem';
import { connect } from 'react-redux';
import SelectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-desktop">Expense</div>
            <div className="show-for-desktop">Amount</div>
        </div>
        <div className="list-body">
            {
                props.expenses.length === 0 ?
                    (<p className="list-item list-item__message">No expenses</p>) :
                    (props.expenses.map((expense) => (<ExpenseListItem key={expense.id} {...expense} />)))
            }
        </div>

    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: SelectExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseList);
