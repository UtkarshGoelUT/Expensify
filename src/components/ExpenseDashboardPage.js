import React from 'react';
import ExpenseList from './ExpenseList';
import ExpeneseListFilters from './ExpenseListFilter';
import ExpensesSummary from './ExpensesSummary';

const ExpenseDashboardPage = () => (
    <div>
        <ExpensesSummary />
        <ExpeneseListFilters />
        <ExpenseList />
    </div>
);

export default ExpenseDashboardPage;