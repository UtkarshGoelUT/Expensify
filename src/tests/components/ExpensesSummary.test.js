import React from 'react';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import moment from 'moment';

const filters = {
    text: '',
    sortBy: 'date',
    startDate: moment().startOf('month'),
    endDate: moment().endOf('month')

}

test("Should render with 1 expense summary", () => {
    const wrapper = shallow(<ExpensesSummary expenses={[expenses[0]]} filters={filters} />);
    expect(wrapper).toMatchSnapshot();
});

test("Should render with expenses summary", () => {
    const wrapper = shallow(<ExpensesSummary expenses={expenses} filters={filters} />);
    expect(wrapper).toMatchSnapshot();
});

