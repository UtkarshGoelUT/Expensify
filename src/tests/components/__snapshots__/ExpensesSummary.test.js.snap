// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render with 1 expense summary 1`] = `
<div>
  Viewing 
  1
   expense : Totalling ₹
  1.95
</div>
`;

exports[`Should render with expenses summary 1`] = `
<div>
  Viewing 
  2
   expenses : Totalling ₹
  46.95
</div>
`;
